text/0089-tenant-token.md

```diff
@@ -0,0 +1,125 @@
+- `Tenant tokens` cannot be less restrictive than their respective parent `API key` and can only be used for searching with a predefined forced filter field.
+- `Tenant tokens` are not stored and thus not retrievable on MeiliSearch. This is why we highly advise setting an expiration time on them for security reasons.
+
+#### Solving Multi-Tenancy with `Tenant tokens`
+
+![](https://i.imgur.com/J4jVe1n.png)
+
+Let's say that `Mark` is a developer for a SaaS platform. He would like to ensure that an end-user can only access their documents at search time. **His database contains many users and he hopes to have many more in the future.**
+
+When a user registers, the backend code generates a `Tenant token` specifically for that end-user so they can only access their documents.
+
+The `filter` parameter is set to restrict the search for documents having an `user_id` attribute. This `filter` parameter is contained in the `Tenant token` payload and cannot be modified during the search by the end-user making the request. `filter` can be made of any valid filters. e.g. `user_id = 10 and category = Romantic`
+
+This `Tenant token` is generated from a parent `API key` used to cipher the `Tenant token`. On the MeiliSearch side, the payload of the `tenant token` defines what data the user is allowed to retrieve, in order words, MeiliSearch grants authorization based on the permission defined in the Tenant token payload.
+
+---
+
+#### Generating a `Tenant token`
+
+```javascript
+const tenantTokenRestrictions = {
+     "indexesPolicy": {
+         "products": {
+             "filter": "user_id = 1"
+         },
+         "reviews": {
+             "filter": "user_id = 1 AND published = true"
+         }
+     },
+    "expiresIn": 3600
+}
+
+export const generateTenantToken = () => {
+  return (parentApiKey: string, restrictions: tenantTokenRestrictions): string => {
+    //extract the 8 first characters of the parentApiKey
+    const prefix = parentApiKey.substring(0,8);
+
+    //serialize restrictions (indexesPolicies object and expiresIn)
+    const queryParameters = serializeQueryParameters(restrictions);
+
+    //create the secured part (parentApiKey + queryParameters)
+    const securedKey = crypto
+      .createHmac('sha256', parentApiKey)
+      .update(queryParameters)
+      .digest('hex');
+
+    //return the generated `Tenant Token`
+    return Buffer.from(prefix + securedKey + queryParameters).toString('base64');
+  };
+};
+```
+
+##### tenantTokenRestrictions
+
+The format allows defining specific enforced search filters for accessible indexes (these indexes must be defined in the parent `Api Key` used to generate the `Tenant Token` and have the search action).
+
+If the user does not want to define specific filters for each index accessible to the end-user, he can use the `*` index wildcard rule.
```

Comment:
Can we add explicitly that this gives access to all indexes and to all documents only in search? If it's the case.

Reply:
It does not give access to every index on the MeiliSearch instance but defines this specific filter for every index accessible for the parent API key used to generate the tenant token.

I'll clarify that!

Reply:
Yeah, sorry I thought that `*` gives access to everything.

Reply:
No problem, the specification must be more precise.